funnel/views/participant.py

```diff
@@ -0,0 +1,164 @@
+    (Participant, {'id': 'participant_id'}, 'participant'),
+    permission='participant-edit')
+def participant_edit(profile, space, participant):
+    form = ParticipantForm(obj=participant, model=Participant)
+    form.events.query = space.events
+    if form.validate_on_submit():
+        form.populate_obj(participant)
+        db.session.commit()
+        flash(_(u"Your changes have been saved"), 'info')
+        return redirect(space.url_for('admin'), code=303)
+    return render_form(form=form, title=_(u"Edit Participant"), submit=_(u"Save changes"))
+
+
+@app.route('/<space>/participant', subdomain='<profile>')
+@lastuser.requires_login
+@load_models(
+    (Profile, {'name': 'profile'}, 'g.profile'),
+    ((ProposalSpace, ProposalSpaceRedirect), {'name': 'space', 'profile': 'profile'}, 'space'),
+    permission='view')
+def participant(profile, space):
+    participant = Participant.query.filter_by(puk=request.args.get('puk')).first()
+    if not participant:
+        return jsonp(message=u"Not found", code=404)
+    elif participant.key == request.args.get('key'):
+        try:
+            contact_exchange = ContactExchange(user_id=g.user.id, participant_id=participant.id, proposal_space_id=space.id)
+            db.session.add(contact_exchange)
+            db.session.commit()
+        except IntegrityError:
+            app.logger.warning(u"Contact Exchange already present")
+            db.session.rollback()
+        return jsonp(participant=participant_data(participant, space.id, full=True))
+    else:
+        return jsonp(message=u"Unauthorized", code=401)
+
+
+@app.route('/<space>/participant/<participant_id>/badge', subdomain='<profile>')
+@lastuser.requires_login
+@load_models(
+    (Profile, {'name': 'profile'}, 'g.profile'),
+    ((ProposalSpace, ProposalSpaceRedirect), {'name': 'space', 'profile': 'profile'}, 'space'),
+    (Participant, {'id': 'participant_id'}, 'participant'),
+    permission='participant-view')
+def participant_badge(profile, space, participant):
+    return render_template('badge.html', badges=participant_badge_data([participant], space))
+
+
+@app.route('/<space>/event/<name>/checkin/<participant_id>', methods=['POST'], subdomain='<profile>')
+@lastuser.requires_login
+@load_models(
+    (Profile, {'name': 'profile'}, 'g.profile'),
+    ((ProposalSpace, ProposalSpaceRedirect), {'name': 'space', 'profile': 'profile'}, 'space'),
+    (Event, {'name': 'name', 'proposal_space': 'space'}, 'event'),
+    (Participant, {'id': 'participant_id'}, 'participant'),
+    permission='event-checkin')
+def event_checkin(profile, space, event, participant):
+    attendee = Attendee.get(event, participant)
+    form = forms.Form()
+    if form.validate_on_submit():
+        # Toggle check-in status
+        attendee.checked_in = not attendee.checked_in
+        db.session.commit()
+    return redirect(url_for('event', profile=space.profile.name, space=space.name, name=event.name), code=303)
```

Comment:
This function looks like it should (a) return current status for a JSON request and (b) redirect for a HTML request. (`render_with` can't do this yet).